### PR TITLE
Add rules for automatic differentiation

### PR DESCRIPTION
**Pull request by sethaxen**
This PR adds rules for reverse-mode automatic differentiation (AD). Specifically, it makes the forward and reverse transforms and the integration functions differentiable with Zygote.jl.

This is something I need for my project, and ZygoteRules.jl is a very light dependency (load time increases around 0.2s on my machine with this PR) intended to be used for specifying rules without making Zygote a dependency.

With this PR, you can do things like
```julia
using Hankel, Zygote
q = QDHT(30.0, 100)
fr = exp.(-q.r.^2 ./ 2)
Zygote.gradient(r -> integrateR(exp.(-r.^2 ./ 2), q), q.r)[1]
```
which computes the gradient of the integral wrt the gridpoints.

At some point in the future, the same rules can be migrated to use ChainRulesCore.jl's interface, which will eventually make these rules available to many more AD packages.

### CODE REVIEW

**Comment by chrisbrahms** [comment]
Sorry for the slow response--had a conference deadline on Friday.

I can't pretend to fully understand how this works since I don't use automatic differentiation myself, but in principle I'm happy to have this. Two things:
1. something has gone wrong with the CI testing. I've tried rerunning the CI but it still won't run. Not sure what's going on there--any ideas?
2. this should be documented, ideally both in the docs (with a quick example) and the readme.

**Comment by sethaxen** [comment]
> Sorry for the slow response--had a conference deadline on Friday.

No worries!

> I can't pretend to fully understand how this works since I don't use automatic differentiation myself, but in principle I'm happy to have this.

Great! I'm happy to explain more if you'd like. Autodiff rules are bit nonintuitive at first but then click. The [ChainRules.jl docs](http://www.juliadiff.org/ChainRulesCore.jl/stable/#The-propagators:-pushforward-and-pullback-1) are also nice.

> Two things:
> 
> 1. something has gone wrong with the CI testing. I've tried rerunning the CI but it still won't run. Not sure what's going on there--any ideas?

I don't know. Looks good to me now.

> 2. this should be documented, ideally both in the docs (with a quick example) and the readme.

Absolutely. I'll add an example as well. I switched over to ChainRules since https://github.com/FluxML/Zygote.jl/pull/366, which makes Zygote use ChainRules' rules, is almost complete. So we can hold off merging this PR until that one has been merged.

I also refactored `symmetric`, `Rsymmetric`, and `oversample` to make them autodiffable without any additional rules (basically, use `cat` and `view` instead of mutating). Now also if one wants to oversample many times, the larger `QDHT` can be pre-calculated and used. Now all of the exported functions are compatible with autodiff. Behavior should be the same. I still need to add tests for the forward-mode differentiation rules.

**Comment by sethaxen** [comment]
Okay, this should be feature-complete now, along with documentation. It's only waiting on final review and for https://github.com/FluxML/Zygote.jl/pull/366 to be merged so that the example in the docs works.

**Comment by sethaxen** [comment]
@chrisbrahms Zygote's master branch now uses ChainRules, so the example in the docs works. This is ready for final review.

File: .github/workflows/documenter.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
         with:
           version: 1.4
       - name: Install dependencies
-        run: julia --project=docs/ -e 'using Pkg; Pkg.develop(PackageSpec(path=pwd())); Pkg.instantiate()'
+        run: julia --project=docs/ -e 'using Pkg; Pkg.instantiate(); Pkg.develop(PackageSpec(path=pwd()))'
```
Comment (chrisbrahms) [inline .github/workflows/documenter.yml:19]: What's the reason for this?

File: .gitignore [hunk 2]
```diff
@@ -4,5 +4,5 @@
 deps/deps.jl
 *.mem
 *.h5
-Manifest.toml
+/Manifest.toml
```
Comment (chrisbrahms) [inline .gitignore:7]: Why do we want to check in the docs `Manifest.toml`?

**Review by chrisbrahms — COMMENTED** [review]
Looks good to me. Just a bit confused by the changes to `.gitignore` and the documenter workflow.

File: .gitignore (hunk 2 above)
Comment (sethaxen) [reply .gitignore:7]: Currently only Zygote#master uses ChainRules, so we need to check in the manifest until a release is made (Zygote's release schedule is typically a bit slower, so I'd rather not wait). When Zygote does make a release, I'll open a new PR removing the docs manifest.

File: .github/workflows/documenter.yml (hunk 1 above)
Comment (sethaxen) [reply .github/workflows/documenter.yml:19]: I didn't see this documented anywhere, but the original order does not work. It's apparently necessary to do things in this order when using an unregistered package in the docs.

Comment (chrisbrahms) [reply .github/workflows/documenter.yml:19]: Yes that makes sense. In the previous order, `Pkg` would not use the `Manifest.toml` and install the latest release of `Zygote` instead.

**Pull request merged**